Parser should reject wildcards in invalid positions

Seems to be caused by wildcards in unions or intersections in type bounds.

```scala
class X[A >: _ | X[_]]
// or
class X[A >: _ | _]
// or
class X[A <: (Int & _)]
```

```scala
Exception in thread "main" java.lang.AssertionError: assertion failed: TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any)) AppliedType(TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,module class <empty>)),module class App$)),class X),List(TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any))))
	at scala.Predef$.assert(Predef.scala:219)
	at dotty.tools.dotc.core.Types$OrType.<init>(Types.scala:2523)
	at dotty.tools.dotc.core.Types$CachedOrType.<init>(Types.scala:2551)
	at dotty.tools.dotc.core.Types$OrType$.apply(Types.scala:2556)
	at dotty.tools.dotc.typer.TypeAssigner.assignType(TypeAssigner.scala:488)
	at dotty.tools.dotc.typer.TypeAssigner.assignType$(TypeAssigner.scala:487)
	at dotty.tools.dotc.typer.Typer.assignType(Typer.scala:83)
	at dotty.tools.dotc.typer.Typer.$anonfun$typedOrTypeTree$1(Typer.scala:1188)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedOrTypeTree(Typer.scala:1184)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1801)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1829)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1816)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1829)
	at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1860)
	at dotty.tools.dotc.reporting.trace$.op1$3(trace.scala:32)
	at dotty.tools.dotc.reporting.trace$.apply(trace.scala:33)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1856)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1868)
	at dotty.tools.dotc.typer.Typer.$anonfun$typedTypeBoundsTree$1(Typer.scala:1286)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedTypeBoundsTree(Typer.scala:1284)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1806)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1829)
	at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1860)
	at dotty.tools.dotc.reporting.trace$.op1$3(trace.scala:32)
	at dotty.tools.dotc.reporting.trace$.apply(trace.scala:33)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1856)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1868)
	at dotty.tools.dotc.typer.Namer.$anonfun$typedAheadType$1(Namer.scala:1006)
	at dotty.tools.dotc.typer.Namer.typedAheadImpl(Namer.scala:999)
	at dotty.tools.dotc.typer.Namer.typedAheadType(Namer.scala:1006)
	at dotty.tools.dotc.typer.Namer.typeDefSig(Namer.scala:1256)
	at dotty.tools.dotc.typer.Namer$TypeDefCompleter.typeSig(Namer.scala:878)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:843)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:821)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:248)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeInfo$1(SymDenotations.scala:211)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:213)
	at dotty.tools.dotc.ast.desugar$DerivedFromParamTree.derivedTree(Desugar.scala:98)
	at dotty.tools.dotc.ast.desugar$DerivedFromParamTree.derivedTree(Desugar.scala:51)
	at dotty.tools.dotc.typer.Typer.$anonfun$typedTypeTree$1(Typer.scala:1158)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedTypeTree(Typer.scala:1153)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1798)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1829)
	at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1860)
	at dotty.tools.dotc.reporting.trace$.op1$3(trace.scala:32)
	at dotty.tools.dotc.reporting.trace$.apply(trace.scala:33)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1856)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1868)
	at dotty.tools.dotc.typer.Namer.$anonfun$typedAheadType$1(Namer.scala:1006)
	at dotty.tools.dotc.typer.Namer.typedAheadImpl(Namer.scala:999)
	at dotty.tools.dotc.typer.Namer.typedAheadType(Namer.scala:1006)
	at dotty.tools.dotc.typer.Namer.typeDefSig(Namer.scala:1256)
	at dotty.tools.dotc.typer.Namer$TypeDefCompleter.typeSig(Namer.scala:878)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:843)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:821)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:248)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeInfo$1(SymDenotations.scala:211)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:213)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:354)
	at dotty.tools.dotc.typer.Namer.$anonfun$defDefSig$1(Namer.scala:1217)
	at dotty.tools.dotc.typer.Namer.$anonfun$defDefSig$1$adapted(Namer.scala:1217)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at dotty.tools.dotc.typer.Namer.defDefSig(Namer.scala:1217)
	at dotty.tools.dotc.typer.Namer$Completer.typeSig(Namer.scala:796)
	at dotty.tools.dotc.typer.Namer$Completer.completeInCreationContext(Namer.scala:843)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:821)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:248)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeInfo$1(SymDenotations.scala:211)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:213)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:354)
	at dotty.tools.dotc.typer.Namer$ClassCompleter.completeInCreationContext(Namer.scala:979)
	at dotty.tools.dotc.typer.Namer$Completer.complete(Namer.scala:821)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeFrom(SymDenotations.scala:248)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.completeInfo$1(SymDenotations.scala:211)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.info(SymDenotations.scala:213)
	at dotty.tools.dotc.core.SymDenotations$SymDenotation.ensureCompleted(SymDenotations.scala:354)
	at dotty.tools.dotc.typer.Typer.retrieveSym(Typer.scala:1730)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1755)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1828)
	at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1860)
	at dotty.tools.dotc.reporting.trace$.op1$3(trace.scala:32)
	at dotty.tools.dotc.reporting.trace$.apply(trace.scala:33)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1856)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1868)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1887)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1916)
	at dotty.tools.dotc.typer.Typer.$anonfun$typedClassDef$1(Typer.scala:1529)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1455)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1768)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1828)
	at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1860)
	at dotty.tools.dotc.reporting.trace$.op1$3(trace.scala:32)
	at dotty.tools.dotc.reporting.trace$.apply(trace.scala:33)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1856)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1868)
	at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:1887)
	at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1916)
	at dotty.tools.dotc.typer.Typer.$anonfun$typedPackageDef$1(Typer.scala:1636)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:1627)
	at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:1808)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1829)
	at dotty.tools.dotc.typer.Typer.$anonfun$typed$2(Typer.scala:1860)
	at dotty.tools.dotc.reporting.trace$.op1$3(trace.scala:32)
	at dotty.tools.dotc.reporting.trace$.apply(trace.scala:33)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1856)
	at dotty.tools.dotc.typer.Typer.typed(Typer.scala:1868)
	at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1928)
	at dotty.tools.dotc.typer.FrontEnd.$anonfun$typeCheck$1(FrontEnd.scala:66)
	at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:34)
	at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:64)
	at dotty.tools.dotc.typer.FrontEnd.$anonfun$runOn$7(FrontEnd.scala:95)
	at dotty.tools.dotc.typer.FrontEnd.$anonfun$runOn$7$adapted(FrontEnd.scala:95)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:95)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$3(Run.scala:174)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.trackTime(Stats.scala:47)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$2(Run.scala:171)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$2$adapted(Run.scala:169)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
	at dotty.tools.dotc.Run.runPhases$1(Run.scala:169)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$1(Run.scala:194)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:88)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:149)
	at dotty.tools.dotc.Run.compileSources(Run.scala:136)
	at dotty.tools.dotc.Run.compile(Run.scala:120)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:29)
	at dotty.tools.dotc.Driver.process(Driver.scala:127)
	at dotty.tools.dotc.Driver.process(Driver.scala:96)
	at dotty.tools.dotc.Driver.process(Driver.scala:108)
	at dotty.tools.dotc.Driver.main(Driver.scala:135)
	at dotty.tools.dotc.Main.main(Main.scala)
```

--- a/compiler/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/compiler/src/dotty/tools/dotc/ast/Desugar.scala
@@ -1130,9 +1130,7 @@ object desugar {
         Apply(Select(Apply(Ident(nme.StringContext), strs), id), elems)
       case InfixOp(l, op, r) =>
         if (ctx.mode is Mode.Type)
-          if (!op.isBackquoted && op.name == tpnme.raw.AMP) AndTypeTree(l, r)     // l & r
-          else if (!op.isBackquoted && op.name == tpnme.raw.BAR) OrTypeTree(l, r) // l | r
-          else AppliedTypeTree(op, l :: r :: Nil) // op[l, r]
+          AppliedTypeTree(op, l :: r :: Nil) // op[l, r]
         else {
           assert(ctx.mode is Mode.Pattern) // expressions are handled separately by `binop`
           Apply(op, l :: r :: Nil) // op(l, r)

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -1008,20 +1008,20 @@ object Parsers {
       else if (location == Location.InPattern) refinedType()
       else infixType()
 
-    /** Checks whether `t` is a wildcard type.
-     *  If it is, returns the [[Position]] where the wildcard occurs.
+    /** Checks whether `t` represents a non-value type (wildcard types, or ByNameTypeTree).
+     *  If it is, returns the [[Tree]] which immediately represents the non-value type.
      */
     @tailrec
-    private final def findWildcardType(t: Tree, alsoNonValue: Boolean): Option[Tree] = t match {
+    private final def findNonValueTypeTree(t: Tree, alsoNonValue: Boolean): Option[Tree] = t match {
       case TypeBoundsTree(_, _) => Some(t)
       case ByNameTypeTree(_) if alsoNonValue => Some(t)
-      case Parens(t1) => findWildcardType(t1, alsoNonValue)
-      case Annotated(t1, _) => findWildcardType(t1, alsoNonValue)
+      case Parens(t1) => findNonValueTypeTree(t1, alsoNonValue)
+      case Annotated(t1, _) => findNonValueTypeTree(t1, alsoNonValue)
       case _ => None
     }
 
     def rejectWildcard(t: Tree, fallbackTree: Tree): Tree =
-      findWildcardType(t, false) match {
+      findNonValueTypeTree(t, false) match {
         case Some(wildcardTree) =>
           syntaxError(UnboundWildcardType(), wildcardTree.pos)
           fallbackTree
@@ -1036,7 +1036,7 @@ object Parsers {
         rejectWildcard(t, fallbackTree)
 
     def checkAndOrArgument(t: Tree): Tree =
-      findWildcardType(t, true) match {
+      findNonValueTypeTree(t, true) match {
         case Some(typTree) =>
           typTree match {
             case typTree: TypeBoundsTree =>

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -2287,7 +2287,8 @@ object Parsers {
     /** ConstrApp         ::=  SimpleType {ParArgumentExprs}
      */
     val constrApp = () => {
-      val t = checkWildcard(annotType())
+      // Using Ident(nme.ERROR) to avoid causing cascade errors on non-user-written code
+      val t = checkWildcard(annotType(), fallbackTree = Ident(nme.ERROR))
       if (in.token == LPAREN) parArgumentExprss(wrapNew(t))
       else t
     }

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -835,7 +835,7 @@ object Parsers {
         if (ctx.settings.strict.value)
           deprecationWarning(DeprecatedWithOperator())
         in.nextToken()
-        AndTypeTree(t, withType())
+        AndTypeTree(checkWildcard(t), checkWildcard(withType()))
       }
       else t
 

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -452,7 +452,7 @@ object Parsers {
       if (isLeftAssoc(op1) != op2LeftAssoc)
         syntaxError(MixedLeftAndRightAssociativeOps(op1, op2, op2LeftAssoc), offset)
 
-    def reduceStack(base: List[OpInfo], top: Tree, prec: Int, leftAssoc: Boolean, op2: Name): Tree = {
+    def reduceStack(base: List[OpInfo], top: Tree, prec: Int, leftAssoc: Boolean, op2: Name, isType: Boolean): Tree = {
       if (opStack != base && precedence(opStack.head.operator.name) == prec)
         checkAssoc(opStack.head.offset, opStack.head.operator.name, op2, leftAssoc)
       def recur(top: Tree): Tree = {
@@ -464,7 +464,15 @@ object Parsers {
             opStack = opStack.tail
             recur {
               atPos(opInfo.operator.pos union opInfo.operand.pos union top.pos) {
-                InfixOp(opInfo.operand, opInfo.operator, top)
+                val op = opInfo.operator
+                val l = opInfo.operand
+                val r = top
+                if (isType && !op.isBackquoted && op.name == tpnme.raw.BAR) {
+                  OrTypeTree(l, r)
+                } else if (isType && !op.isBackquoted && op.name == tpnme.raw.AMP) {
+                  AndTypeTree(l, r)
+                } else
+                  InfixOp(l, op, r)
               }
             }
           }
@@ -488,20 +496,20 @@ object Parsers {
       var top = first
       while (isIdent && isOperator) {
         val op = if (isType) typeIdent() else termIdent()
-        top = reduceStack(base, top, precedence(op.name), isLeftAssoc(op.name), op.name)
+        top = reduceStack(base, top, precedence(op.name), isLeftAssoc(op.name), op.name, isType)
         opStack = OpInfo(top, op, in.offset) :: opStack
         newLineOptWhenFollowing(canStartOperand)
         if (maybePostfix && !canStartOperand(in.token)) {
           val topInfo = opStack.head
           opStack = opStack.tail
-          val od = reduceStack(base, topInfo.operand, 0, true, in.name)
+          val od = reduceStack(base, topInfo.operand, 0, true, in.name, isType)
           return atPos(startOffset(od), topInfo.offset) {
             PostfixOp(od, topInfo.operator)
           }
         }
         top = operand()
       }
-      reduceStack(base, top, 0, true, in.name)
+      reduceStack(base, top, 0, true, in.name, isType)
     }
 
 /* -------- IDENTIFIERS AND LITERALS ------------------------------------------- */

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -468,9 +468,9 @@ object Parsers {
                 val l = opInfo.operand
                 val r = top
                 if (isType && !op.isBackquoted && op.name == tpnme.raw.BAR) {
-                  OrTypeTree(l, r)
+                  OrTypeTree(checkWildcard(l), checkWildcard(r))
                 } else if (isType && !op.isBackquoted && op.name == tpnme.raw.AMP) {
-                  AndTypeTree(l, r)
+                  AndTypeTree(checkWildcard(l), checkWildcard(r))
                 } else
                   InfixOp(l, op, r)
               }
@@ -784,7 +784,7 @@ object Parsers {
           val start = in.offset
           val tparams = typeParamClause(ParamOwner.TypeParam)
           if (in.token == ARROW)
-            atPos(start, in.skipToken())(LambdaTypeTree(tparams, typ()))
+            atPos(start, in.skipToken())(LambdaTypeTree(tparams, toplevelTyp()))
           else { accept(ARROW); typ() }
         }
         else infixType()
@@ -822,7 +822,7 @@ object Parsers {
 
     def refinedTypeRest(t: Tree): Tree = {
       newLineOptWhenFollowedBy(LBRACE)
-      if (in.token == LBRACE) refinedTypeRest(atPos(startOffset(t)) { RefinedTypeTree(t, refinement()) })
+      if (in.token == LBRACE) refinedTypeRest(atPos(startOffset(t)) { RefinedTypeTree(checkWildcard(t), refinement()) })
       else t
     }
 
@@ -886,7 +886,7 @@ object Parsers {
     private def simpleTypeRest(t: Tree): Tree = in.token match {
       case HASH => simpleTypeRest(typeProjection(t))
       case LBRACKET => simpleTypeRest(atPos(startOffset(t)) {
-        AppliedTypeTree(t, typeArgs(namedOK = false, wildOK = true)) })
+        AppliedTypeTree(checkWildcard(t), typeArgs(namedOK = false, wildOK = true)) })
       case _ => t
     }
 
@@ -2159,7 +2159,7 @@ object Parsers {
         in.token match {
           case EQUALS =>
             in.nextToken()
-            TypeDef(name, lambdaAbstract(tparams, typ())).withMods(mods).setComment(in.getDocComment(start))
+            TypeDef(name, lambdaAbstract(tparams, toplevelTyp())).withMods(mods).setComment(in.getDocComment(start))
           case SUPERTYPE | SUBTYPE | SEMI | NEWLINE | NEWLINES | COMMA | RBRACE | EOF =>
             TypeDef(name, lambdaAbstract(tparams, typeBounds())).withMods(mods).setComment(in.getDocComment(start))
           case _ =>
@@ -2287,7 +2287,7 @@ object Parsers {
     /** ConstrApp         ::=  SimpleType {ParArgumentExprs}
      */
     val constrApp = () => {
-      val t = annotType()
+      val t = checkWildcard(annotType())
       if (in.token == LPAREN) parArgumentExprss(wrapNew(t))
       else t
     }

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -768,9 +768,16 @@ object Parsers {
             accept(RPAREN)
             if (imods.is(Implicit) || isValParamList || in.token == ARROW) functionRest(ts)
             else {
-              for (t <- ts)
-                if (t.isInstanceOf[ByNameTypeTree])
-                  syntaxError(ByNameParameterNotSupported())
+              val ts1 =
+                for (t <- ts) yield {
+                  t match {
+                    case t@ByNameTypeTree(t1) =>
+                      syntaxError(ByNameParameterNotSupported(t), t.pos)
+                      t1
+                    case _ =>
+                      t
+                  }
+                }
               val tuple = atPos(start) { makeTupleOrParens(ts) }
               infixTypeRest(
                 refinedTypeRest(

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -468,9 +468,9 @@ object Parsers {
                 val l = opInfo.operand
                 val r = top
                 if (isType && !op.isBackquoted && op.name == tpnme.raw.BAR) {
-                  OrTypeTree(checkWildcard(l), checkWildcard(r))
+                  OrTypeTree(checkAndOrArgument(l), checkAndOrArgument(r))
                 } else if (isType && !op.isBackquoted && op.name == tpnme.raw.AMP) {
-                  AndTypeTree(checkWildcard(l), checkWildcard(r))
+                  AndTypeTree(checkAndOrArgument(l), checkAndOrArgument(r))
                 } else
                   InfixOp(l, op, r)
               }
@@ -842,7 +842,7 @@ object Parsers {
         if (ctx.settings.strict.value)
           deprecationWarning(DeprecatedWithOperator())
         in.nextToken()
-        AndTypeTree(checkWildcard(t), checkWildcard(withType()))
+        AndTypeTree(checkAndOrArgument(t), checkAndOrArgument(withType()))
       }
       else t
 
@@ -1034,6 +1034,19 @@ object Parsers {
         t
       else
         rejectWildcard(t, fallbackTree)
+
+    def checkAndOrArgument(t: Tree): Tree =
+      findWildcardType(t, true) match {
+        case Some(typTree) =>
+          typTree match {
+            case typTree: TypeBoundsTree =>
+              syntaxError(UnboundWildcardType(), typTree.pos)
+            case typTree: ByNameTypeTree =>
+              syntaxError(ByNameParameterNotSupported(typTree), typTree.pos)
+          }
+          scalaAny
+        case None => t
+      }
 
 /* ----------- EXPRESSIONS ------------------------------------------------ */
 

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -717,15 +717,7 @@ object Parsers {
     /** Same as [[typ]], but if this results in a wildcard it emits a syntax error and
      *  returns a tree for type `Any` instead.
      */
-    def toplevelTyp(): Tree = {
-      val t = typ()
-      findWildcardType(t) match {
-        case Some(wildcardPos) =>
-          syntaxError(UnboundWildcardType(), wildcardPos)
-          scalaAny
-        case None => t
-      }
-    }
+    def toplevelTyp(): Tree = checkWildcard(typ())
 
     /** Type        ::=  [FunArgMods] FunArgTypes `=>' Type
      *                |  HkTypeParamClause `->' Type
@@ -925,7 +917,7 @@ object Parsers {
           else Nil
         first :: rest
       }
-      def typParser() = if (wildOK) typ() else toplevelTyp()
+      def typParser() = checkWildcard(typ(), wildOK)
       if (namedOK && in.token == IDENTIFIER)
         typParser() match {
           case Ident(name) if in.token == EQUALS =>
@@ -1019,6 +1011,17 @@ object Parsers {
       case Annotated(t1, _) => findWildcardType(t1)
       case _ => None
     }
+
+    def rejectWildcard(t: Tree, report: Position => Unit, fallbackTree: Tree): Tree =
+      findWildcardType(t) match {
+        case Some(wildcardPos) =>
+          report(wildcardPos)
+          fallbackTree
+        case None => t
+      }
+
+    def checkWildcard(t: Tree, wildOK: Boolean = false, fallbackTree: Tree = scalaAny): Tree =
+      if (wildOK) t else rejectWildcard(t, syntaxError(UnboundWildcardType(), _), fallbackTree)
 
 /* ----------- EXPRESSIONS ------------------------------------------------ */
 

--- a/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
+++ b/compiler/src/dotty/tools/dotc/parsing/Parsers.scala
@@ -1012,23 +1012,28 @@ object Parsers {
      *  If it is, returns the [[Position]] where the wildcard occurs.
      */
     @tailrec
-    private final def findWildcardType(t: Tree): Option[Position] = t match {
-      case TypeBoundsTree(_, _) => Some(t.pos)
-      case Parens(t1) => findWildcardType(t1)
-      case Annotated(t1, _) => findWildcardType(t1)
+    private final def findWildcardType(t: Tree, alsoNonValue: Boolean): Option[Tree] = t match {
+      case TypeBoundsTree(_, _) => Some(t)
+      case ByNameTypeTree(_) if alsoNonValue => Some(t)
+      case Parens(t1) => findWildcardType(t1, alsoNonValue)
+      case Annotated(t1, _) => findWildcardType(t1, alsoNonValue)
       case _ => None
     }
 
-    def rejectWildcard(t: Tree, report: Position => Unit, fallbackTree: Tree): Tree =
-      findWildcardType(t) match {
-        case Some(wildcardPos) =>
-          report(wildcardPos)
+    def rejectWildcard(t: Tree, fallbackTree: Tree): Tree =
+      findWildcardType(t, false) match {
+        case Some(wildcardTree) =>
+          syntaxError(UnboundWildcardType(), wildcardTree.pos)
           fallbackTree
         case None => t
       }
 
+
     def checkWildcard(t: Tree, wildOK: Boolean = false, fallbackTree: Tree = scalaAny): Tree =
-      if (wildOK) t else rejectWildcard(t, syntaxError(UnboundWildcardType(), _), fallbackTree)
+      if (wildOK)
+        t
+      else
+        rejectWildcard(t, fallbackTree)
 
 /* ----------- EXPRESSIONS ------------------------------------------------ */
 

--- a/compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala
+++ b/compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala
@@ -696,10 +696,10 @@ object messages {
     }
   }
 
-  case class ByNameParameterNotSupported()(implicit ctx: Context)
+  case class ByNameParameterNotSupported(tpe: untpd.TypTree)(implicit ctx: Context)
   extends Message(ByNameParameterNotSupportedID) {
     val kind = "Syntax"
-    val msg = "By-name parameter type not allowed here."
+    val msg = hl"By-name parameter type ${tpe} not allowed here."
 
     val explanation =
       hl"""|By-name parameters act like functions that are only evaluated when referenced,

--- a/tests/neg/i4373.scala
+++ b/tests/neg/i4373.scala
@@ -1,0 +1,23 @@
+trait Base
+trait TypeConstr[X]
+
+class X1[A >: _ | X1[_]] // error
+class X2[A >: _ & X2[_]] // error
+class X3[A >: X3[_] | _] // error
+class X4[A >: X4[_] & _] // error
+
+class A1 extends _ // error
+class A2 extends _ with _ // error // error
+class A3 extends Base with _ // error
+class A4 extends _ with Base // error
+
+object Test {
+  type T1 = _ // error
+  type T2 = _[Int] // error
+  type T3 = _ { type S } // error
+  type T4 = [X] => _ // error
+
+  // Open questions:
+  type T5 = TypeConstr[_ { type S }] // error
+  type T5 = TypeConstr[_[Int]] // error
+}

--- a/tests/neg/i4373.scala
+++ b/tests/neg/i4373.scala
@@ -30,4 +30,22 @@ object Test {
   type T10 = (Int => Int) | (=> Int) // error
   type T11 = (Int => Int) & (=> Int) // error
   type T12 = (Int => Int) with (=> Int) // error
+
+  // annotations
+  type T13 = _ @ annotation.tailrec // error
+  type T14 = Int @ _ // error
+  type T15 = (_ | Int) @ annotation.tailrec // error
+  type T16 = (Int | _) @ annotation.tailrec // error
+  type T17 = Int @ (_ | annotation.tailrec) // error
+  type T18 = Int @ (annotation.tailrec | _) // error
+
+  type T19 = (_ with Int) @ annotation.tailrec // error
+  type T20 = (Int with _) @ annotation.tailrec // error
+  type T21 = Int @ (_ with annotation.tailrec) // error // error
+  type T22 = Int @ (annotation.tailrec with _) // error // error
+
+  type T23 = (_ & Int) @ annotation.tailrec // error
+  type T24 = (Int & _) @ annotation.tailrec // error
+  type T25 = Int @ (_ & annotation.tailrec) // error
+  type T26 = Int @ (annotation.tailrec & _) // error
 }

--- a/tests/neg/i4373.scala
+++ b/tests/neg/i4373.scala
@@ -21,5 +21,13 @@ object Test {
 
   // Open questions:
   type T5 = TypeConstr[_ { type S }] // error
-  type T5 = TypeConstr[_[Int]] // error
+  type T6 = TypeConstr[_[Int]] // error
+
+  // expression types
+  type T7 = (=> Int) | (Int => Int) // error
+  type T8 = (=> Int) & (Int => Int) // error
+  type T9 = (=> Int) with (Int => Int) // error
+  type T10 = (Int => Int) | (=> Int) // error
+  type T11 = (Int => Int) & (=> Int) // error
+  type T12 = (Int => Int) with (=> Int) // error
 }

--- a/tests/neg/i4373.scala
+++ b/tests/neg/i4373.scala
@@ -5,6 +5,8 @@ class X1[A >: _ | X1[_]] // error
 class X2[A >: _ & X2[_]] // error
 class X3[A >: X3[_] | _] // error
 class X4[A >: X4[_] & _] // error
+class X5[A >: _ with X5[_]] // error
+class X6[A >: X6[_] with _] // error
 
 class A1 extends _ // error
 class A2 extends _ with _ // error // error

--- a/tests/neg/i4373a.scala
+++ b/tests/neg/i4373a.scala
@@ -1,0 +1,3 @@
+// ==> 040fb47fbaf718cecb11a7d51ac5a48bf4f6a1fe.scala <==
+object x0 {
+  val x0 : _ with // error // error // error

--- a/tests/neg/i4373b.scala
+++ b/tests/neg/i4373b.scala
@@ -1,0 +1,4 @@
+// ==> 05bef7805687ba94da37177f7568e3ba7da1f91c.scala <==
+class x0 {
+  x1: // error
+      x0 | _ // error // error

--- a/tests/neg/i4373c.scala
+++ b/tests/neg/i4373c.scala
@@ -1,0 +1,3 @@
+// ==> 18b253a4a89a84c5674165c6fc3efafad535eee3.scala <==
+object x0 {
+  def x1[x2 <:_[ // error // error // error

